### PR TITLE
Support loading a raw PyTorch state-dict checkpoint in Fabric

### LINKED ISSUE #18041: Support loading a raw checkpoint in `Fabric.load()`
### Description & Motivation

Fabric's format for saving checkpoints looks like this:

```py
state = {"model": ..., "optimizer": ...}
fabric.save("file.pt", state)
```
And the resulting checkpoint file simply contains the pickled dictionary with keys "model" etc. This was made so that you can easily make the training script stateful by loading a checkpoint in-place like so:

```py
state = {"model": ..., "optimizer": ...}
fabric.load("file.pt", state)  # updates state contents
```

However, loading a checkpoint this way assumes it has been saved with Fabric before. What if we want to load a raw checkpoint from a model that wasn't pretrained with Fabric (e.g. lit-gpt)? 

### Pitch

Support loading a raw state dict directly into the model:

```py
model = MyModel()
fabric.load("file.pt", model)  # loads parameters of model
```

This way, the user doesn't need to convert their checkpoints to Fabric's format. The drawback is that this will likely make the `Fabric.load()` and `Strategy.load_checkpoint` logic more complicated as it now needs to handle multiple types of inputs at the same time.

### Alternatives

Advise users to convert their raw checkpoints to the Fabric format. However, this is inconvenient or even impossible for very large model checkpoints. 

### Additional context

This came up in #18008

cc @borda @awaelchli @carmocca @justusschock

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18041 
Asked by users: https://github.com/Lightning-AI/lightning/issues/18008

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @awaelchli @carmocca @justusschock

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -745,6 +749,18 @@ def load(
...
         return remainder
 
+    # Name proposals:
+    # load_raw
+    # load_object
+    # load_torch
+    # load_from_torch
+    # load_non_fabric
+    # load_state_dict (?)
+    def load_object(self, path, obj: Any):
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:759]: Proposed alternative names:
load_raw
load_object
load_torch
load_from_torch
load_non_fabric

Personal opinion is that this feature should be combined with `fabric.load()`.

File: src/lightning/fabric/strategies/fsdp.py [hunk 2]
```diff
@@ -459,19 +459,28 @@ def load_checkpoint(
...
         path = Path(self.broadcast(path))
 
+        if isinstance(state, Module):
+            _load_raw_module_state(path, module=state, strict=strict)
+            return {}
+
+        if isinstance(state, Optimizer):
+            raise NotImplementedError(
+                "Loading a single optimizer object from a checkpoint is not supported yet with the FSDP strategy."
+            )
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:469]: todo test

File: src/lightning/fabric/strategies/deepspeed.py [hunk 3]
```diff
@@ -467,7 +467,7 @@ def load_checkpoint(
                 If DeepSpeed was unable to load the checkpoint due to missing files or because the checkpoint is
                 not in the expected DeepSpeed format.
         """
-        if self.load_full_weights and self.zero_stage_3:
+        if isinstance(state, (Module, Optimizer)) or self.load_full_weights and self.zero_stage_3:
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/deepspeed.py:470]: todo  test

File: src/lightning/fabric/fabric.py [hunk 4]
```diff
@@ -745,6 +745,22 @@ def load(
                 state[k] = unwrapped_state[k]
         return remainder
 
+    def load_object(self, path: Union[str, Path], obj: Union[nn.Module, Optimizer], strict: bool = True) -> None:
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:748]: docs

File: src/lightning/fabric/strategies/fsdp.py [hunk 5]
```diff
@@ -683,6 +692,24 @@ def _is_full_checkpoint(path: Path) -> bool:
...
 
 
+def _load_raw_module_state(path_or_ckpt: Union[Path, Dict[str, Any]], module: Module, strict: bool = None) -> None:
+    """Loads the state dict (given or from a path) into the module by gathering all weights first and then and
+    writing back to each shard."""
+    if isinstance(path_or_ckpt, Path) and not _is_full_checkpoint(path_or_ckpt):
+        raise ValueError(
+            "Failed to load checkpoint directly into the model. The given path must be a single file containing the"
+            f" full state dict: {path_or_ckpt}"
+        )
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:702]: todo test

File: src/lightning/fabric/fabric.py [hunk 6]
```diff
@@ -746,6 +746,22 @@ def load(
                 state[k] = unwrapped_state[k]
         return remainder
 
+    def load_raw(self, path: Union[str, Path], obj: Union[nn.Module, Optimizer], strict: bool = True) -> None:
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:749]: Point of discussion:

1. Should this be a separate method or the functionality combined in Fabric.load?
2. Alternative names considered: load_raw, load_object, load_torch, load_from_torch

Comment (carmocca) [reply src/lightning/fabric/fabric.py:749]: (1) I like the current approach (i'm biased though)
(2) I prefer `_torch` a little bit - raw could be interpreted as bytes or something similar.

File: src/lightning/fabric/strategies/strategy.py [hunk 7]
```diff
@@ -288,15 +288,19 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
 
         Args:
             path: A path to where the file is located
-            state: A dictionary of objects whose state will be restored in-place from the checkpoint path.
-                If no state is given, then the checkpoint will be returned in full.
+            state: Can be one of:
+
+                - A dictionary of objects whose state will be restored in-place from the checkpoint path.
+                - ``None`` or the empty dict: The loaded checkpoint will be returned in full.
+                - A :class:`~torch.nn.Module` instance, if the checkpoint file contains a raw module state dict.
```
Comment (carmocca) [inline src/lightning/fabric/strategies/strategy.py:302]: Mention optimizer too?

File: tests/tests_fabric/strategies/test_fsdp_integration.py [hunk 8]
```diff
@@ -259,6 +266,26 @@ def test_fsdp_load_full_state_dict_into_sharded_model(tmp_path):
...
+        torch.save(checkpoint["model"], raw_checkpoint_path)
+    fabric.barrier()
+
+    fabric.run()
+    fabric.load_raw(raw_checkpoint_path, fabric.model)
+
+    # Gather all weights and compare
+    with FSDP.summon_full_params(fabric.model, writeback=False, rank0_only=False):
+        params_after = torch.cat([p.cpu().view(-1) for p in fabric.model.parameters()])
+    assert torch.equal(params_before, params_after)
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_fsdp_integration.py:287]: I would find it simpler if we mocked `model._forward_module.load_state_dict` so that we can directly assert what was loaded
Instead of creating copies of the params and asserting that they are equal.

Feel free to ignore if you disagree

**Review by carmocca — APPROVED** [review]
Nice! You have a TODO pending

File: src/lightning/fabric/fabric.py (hunk 6 above)
Comment (justusschock) [reply src/lightning/fabric/fabric.py:749]: I also like the current one.

**Review by justusschock — APPROVED** [review]

File: docs/source-fabric/api/fabric_methods.rst [hunk 9]
```diff
@@ -254,10 +254,30 @@ Fabric will handle the loading part correctly, whether running a single device,
...
+Load the state-dict of a model or optimizer from a raw PyTorch checkpoint not saved by Fabric.
+
+.. code-block:: python
+
+    model = MyModel()
+
+    # A model weights file saved by your friend who doesn't use Fabric
+    fabric.load_raw("path/to/model.pt", model)
+
+    # Equivalent to this:
```
Comment (Borda) [inline docs/source-fabric/api/fabric_methods.rst:277]: maybe have the line below as a comment so not to confuse user if he needs to copy this too?

File: docs/source-fabric/guide/checkpoint.rst [hunk 10]
```diff
@@ -73,6 +76,28 @@ If you want to be in complete control of how states get restored, you can omit p
...
+
+.. code-block:: python
+
+    model = MyModel()
+
+    # A model weights file saved by your friend who doesn't use Fabric
+    fabric.load_raw("path/to/model.pt", model)
+
+    # Equivalent to this:
+    model.load_state_dict(torch.load("path/to/model.pt"))
```
Comment (Borda) [inline docs/source-fabric/guide/checkpoint.rst:92]:
```suggestion
    # model.load_state_dict(torch.load("path/to/model.pt"))
```

File: src/lightning/fabric/strategies/deepspeed.py [hunk 11]
```diff
@@ -444,7 +444,7 @@ def save_checkpoint(
     def load_checkpoint(
         self,
         path: _PATH,
-        state: Optional[Dict[str, Union[Module, Optimizer, Any]]] = None,
+        state: Optional[Union[Module, Optimizer, Dict[str, Union[Module, Optimizer, Any]]]] = None,
```
Comment (Borda) [inline src/lightning/fabric/strategies/deepspeed.py:447]: the typing is getting a it wild lol

**Review by Borda — APPROVED** [review]

File: tests/tests_fabric/strategies/test_fsdp_integration.py (hunk 8 above)
Comment (awaelchli) [reply tests/tests_fabric/strategies/test_fsdp_integration.py:283]: I found that this was the easiest way. Comparing the state dict passed to the mock would be more complicated. Just storing a copy of the flat parameters like I do now doesn't seem so inefficient, and is easy to compare.
